src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlyTensorSpan.cs

```diff
@@ -655,34 +654,63 @@ public ReadOnlyTensorSpan<T> Slice(params scoped ReadOnlySpan<NRange> ranges)
             if (ranges.Length != Lengths.Length)
                 throw new ArgumentOutOfRangeException(nameof(ranges), "Number of dimensions to slice does not equal the number of dimensions in the span");
 
+            ReadOnlyTensorSpan<T> toReturn;
             scoped Span<nint> lengths;
             scoped Span<nint> offsets;
+            nint[]? lengthsArray;
+            nint[]? offsetsArray;
             if (Rank > TensorShape.MaxInlineRank)
             {
-                lengths = stackalloc nint[Rank];
-                offsets = stackalloc nint[Rank];
+                lengthsArray = ArrayPool<nint>.Shared.Rent(Rank);
+                lengths = lengthsArray.AsSpan(0, Rank);
+                lengths.Clear();
+
+                offsetsArray = ArrayPool<nint>.Shared.Rent(Rank);
+                offsets = offsetsArray.AsSpan(0, Rank);
+                offsets.Clear();
             }
             else
             {
-                lengths = new nint[Rank];
-                offsets = new nint[Rank];
+                lengths = stackalloc nint[Rank];
+                offsets = stackalloc nint[Rank];
+
+                lengthsArray = null;
+                offsetsArray = null;
             }
 
             for (int i = 0; i < ranges.Length; i++)
             {
                 (offsets[i], lengths[i]) = ranges[i].GetOffsetAndLength(Lengths[i]);
             }
 
+            // FlattenedLength is computed everytime so using a local to cache the value.
+            nint flattenedLength = FlattenedLength;
             nint index = 0;
             for (int i = 0; i < offsets.Length; i++)
             {
-                index += Strides[i] * (offsets[i]);
+                // When flattenedLength is 0, at least 1 of the Lengths will be 0 and so will the offset which would trigger this exception even though its valid.
+                // To counteract this we specifically check for the case where the offset is 0 and skip the check.
+                if ((nuint)offsets[i] != 0 && (nuint)offsets[i] >= (nuint)Lengths[i])
+                    ThrowHelper.ThrowIndexOutOfRangeException();
+
+                if (lengths[i] > Lengths[i])
+                    ThrowHelper.ThrowIndexOutOfRangeException();
+
+                if (flattenedLength != 0)
+                    index += Strides[i] * (offsets[i]);
```

Comment:
If the above checks were moved up, into the loop that does `GetOffsetAndLength`, we could avoid 3 branches in this loop and instead wrap the loop with `if (flattenedLength != 0)`, reducing cost

Reply:
Fixed.